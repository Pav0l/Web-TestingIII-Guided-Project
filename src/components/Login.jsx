import React from 'react';
import axios from 'axios';

export default class Login extends React.Component {
  state = { flashMessage: '', username: '', password: '' };

  onChange = (field, value) => {
    if (/^[a-z0-9]+$/i.test(value)) {
      this.setState({ [field]: value });
    }
  };

  onLogin = () => {
    axios
      .post('login', {
        username: this.state.username,
        password: this.state.password,
      })
      .then(res => {
        localStorage.setItem('token', res.data.token);
        this.setState({ flashMessage: `Welcome, ${res.data.username}` });
      })
      .catch(res => {
        this.setState({ flashMessage: res.message });
      });
  };

  render() {
    return (
      <div>
        <h3>Login form</h3>
        <div>
          <label htmlFor='usernameInput'>username</label>
          <input
            className='input'
            id='usernameInput'
            value={this.state.username}
            onChange={e => this.onChange('username', e.target.value)}
            placeholder='username'
            type='text'
          />
          <br />
          <label htmlFor='passwordInput'>password</label>
          <input
            className='input'
            id='passwordInput'
            value={this.state.password}
            onChange={e => this.onChange('password', e.target.value)}
            placeholder='password'
            type='text'
          />
        </div>

        {this.state.username && this.state.password ? (
          <button onClick={this.onLogin} data-testid='loginButton'>
            Login
          </button>
        ) : null}
        <div>{this.state.flashMessage}</div>
      </div>
    );
  }
}
