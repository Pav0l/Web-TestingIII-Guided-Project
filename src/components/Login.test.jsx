import React from 'react';
import * as rt from 'react-testing-library';
import Login from './Login';

afterEach(rt.cleanup);

describe('Login', () => {
  it('displays no login button if no username & no password', () => {
    // write test first, see it fail, implement feature, see test pass
    const wrap = rt.render(<Login />);
    const loginBtn = wrap.queryByTestId(/loginButton/i);
    expect(loginBtn).toBeFalsy();
  });

  it('can change input values', () => {
    // we can fire change events on inputs
    // and we can grab inputs by their placeholder texts or their values
    const wrap = rt.render(<Login />);

    /*
      SNAPSHOTING
    */
    // so when running this test, a snapshot of the actuall DOM is taken
    // and saved in __snapshots__ folder inside components folder
    expect(wrap.asFragment()).toMatchSnapshot();
    /*
      â€º 1 snapshot written.
      Snapshot Summary
      â€º 1 snapshot written from 1 test suite.
    */

    // now if you do changes to the DOM, you will get an error when running this code
    // saying that the DOM is different than that of snapshot
    // you can update the snapshot with pressing 'u' if you want to update it
    // to current DOM view

    const inputValue = 'name';
    const passValue = 'secretPassword';

    rt.fireEvent.change(wrap.getByPlaceholderText('username'), {
      target: { value: inputValue },
    });

    // expect(wrap.getByPlaceholderText('username').value).toBe(inputValue);
    expect(wrap.getByDisplayValue(inputValue));

    rt.fireEvent.change(wrap.getByPlaceholderText('password'), {
      target: { value: passValue },
    });
    // expect(wrap.getByPlaceholderText('password').value).toBe(passValue);
    expect(wrap.getByDisplayValue(passValue));
  });

  it('displays login button if username & password', () => {
    const wrap = rt.render(<Login />);
    const usernameInput = wrap.getByLabelText('username');
    const passwordInput = wrap.getByLabelText('password');

    rt.fireEvent.change(usernameInput, { target: { value: 'A' } });

    expect(wrap.queryByTestId(/loginButton/i)).toBeFalsy();

    rt.fireEvent.change(passwordInput, { target: { value: 'b' } });

    expect(wrap.queryByTestId(/loginButton/i)).toBeTruthy();
  });

  it('can login successfully', async () => {
    // see the greeting render
  });

  it('can fail miserably', async () => {
    // see the error render
  });
});
