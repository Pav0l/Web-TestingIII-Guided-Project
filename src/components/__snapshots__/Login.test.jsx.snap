// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login can change input values 1`] = `
<DocumentFragment>
  <div>
    <div>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <div />
  </div>
</DocumentFragment>
`;
